raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1428,51 +1433,66 @@ private Records readCommitted(OffsetAndEpoch offsetAndEpoch, long highWatermark)
     }
 
     @Override
-    public void shutdown(int timeoutMs) {
-        // TODO: Safe to access epoch? Need to reset connections to be able to send EndQuorumEpoch? Block until shutdown completes?
+    public CompletableFuture<Void> shutdown(int timeoutMs) {
+        logger.info("Beginning graceful shutdown");
+        CompletableFuture<Void> shutdownComplete = new CompletableFuture<>();
         kafkaRaftMetrics.close();
-        shutdown.set(new GracefulShutdown(timeoutMs, quorum.epoch()));
+        shutdown.set(new GracefulShutdown(timeoutMs, quorum.epoch(), shutdownComplete));
```

Comment:
Should we close metrics in `GracefulShutdown#complete / update` then?

Reply:
The `Metrics` object is generally shared. It gets shutdown in `RaftServer.shutdown`. Is there any additional cleanup that we need to do here?

Reply:
The `RaftServer.shutdown` is just for initializing the shutdown process and until we call `complete` there maybe end-quorum req/resp still.